### PR TITLE
Fix to overlap logic

### PR DESCRIPTION
**Pull request by hms1**
Implements the changes discussed in [https://github.com/OHDSI/WebAPI/issues/779](url) .

It is important to note that the new overlap logic will recognise a one day overlap between two cohorts where `cohort1.cohort_end_date = cohort2.cohort_start_date`. On its own, this is probably technically correct given the OMOP convention for cohort dates to be inclusive.

However, there is an interaction with the collapse window functionality. If one `cohort_end_date` is within `collapse_window` of another `cohort_start_date`, a one day overlap will be artificially created. While I'm guessing these scenarios are very much edge cases, I doubt it's the behaviour we want. 

[mick-iqvia](/mick-iqvia) and I have been discussing some alternative ways that closely related event windows could be collapsed. We may want to consider incorporating into the same release as this PR.

### CODE REVIEW

**Comment by chrisknoll** [comment]
I'm concerned about the issue you raised:  I understand that certain people in the OHDSI community consider the date diff between 1/1/2001 and 1/1/2001 to be 1 day.  however, that's fur purposes of calculating the duration of a time span assuming the end date includes all 24 hours and the start date does not.  It's this confusion of how to interpret a date is something we've avoided in CIRCE by just treating those dates as 'face value' such that it works out if I add 7 days to 1/1 I get 1/8, and when i subtract 1/8 from 1/1 i get the 7 days.  Treating the 1/8 day as 'special end date handling' means there is actually 8 days between 1/1 and 1/8.

This is all to say:  we should be treating the individual dates as 'face value dates'.  In the case of TxP:  I would not expect the following:

Cohort|start|end|
---|----|----|
A|1/1|1/8
B|1/8|1/12

to result in the pathway:

A -> A+B -> B

It should just be:

A -> B

Like you said, this is only an issue where you have an A that ends exactly when B starts, but with collapse window, you're forcing this to happen to make overlaps not appear.

So, does your change _depend_ on a end date being treated as `date + 1` or can we just leave the current semantics that dates are just dates?

File: src/main/resources/resources/pathway/runPathwayAnalysis.sql [hunk 1]
```diff
@@ -88,62 +88,32 @@ into
...
-  FROM #split_overlapping_events
-)
-SELECT SUM(DISTINCT POWER(2, e.event_cohort_index)) as combo_id, subject_id, cohort_start_date, cohort_end_date
+WITH 
+cohort_dates AS (
+	SELECT DISTINCT subject_id, cohort_date 
+	FROM (
+		  SELECT subject_id, cohort_start_date cohort_date FROM #collapsed_dates_events 
+		  UNION 
+		  SELECT subject_id, DATEADD(day, 1, cohort_end_date) cohort_date FROM #collapsed_dates_events
```
Comment (chrisknoll) [inline src/main/resources/resources/pathway/runPathwayAnalysis.sql]: I see this is where we're padding the end date by 1 day.  is this critical to the algorithm working correctly?  Or is this just a mater of the perspective that an end date includes all hours of the current day?

Comment (hms1) [reply src/main/resources/resources/pathway/runPathwayAnalysis.sql]: The addition of one day is used to define non-overlapping time periods, which is important if we want to follow the general approach suggested. 

It is later on in the query (line 113) where we determine which event cohorts are 'active' during each time period. It'd probably be better to make alterations related to interpretation of cohort start/end date inclusivity/exclusivity here.

**Comment by hms1** [comment]
> Cohort	start	end
> A	1/1	1/8
> B	1/8	1/12
> to result in the pathway:
> 
> A -> A+B -> B
> 
> It should just be:
> 
> A -> B
> 

While I totally agree that A -> B makes more sense intuitively, it is tricky to reconcile a single treatment with other examples. For example, what about the following:

| Cohort  | Start | End |
| ------------- | ------------- | ------------- | 
| A  | 1/1  | 1/1 |
| B  | 1/1  | 1/8 |

Would we not want A + B -> B in this case? 

You could potentially handle these one day cohorts by changing line 113 to:

`WHERE (e.cohort_start_date <= tp.next_cohort_date AND e.cohort_end_date > tp.cohort_date)
OR  (e.cohort_start_date = tp.cohort_date)` (this would need much more careful thought and testing, but you get the idea)

It gets a little messy for my liking, though. The fact that you can have a cohort that starts and ends on the same day makes me think that cohort dates are inclusive (I'm happy to be persuaded otherwise though).

Another approach would be for the SQL to return the most granular pathways as possible, i.e. assume that cohort dates are inclusive, do not collapse any dates and return all overlaps (however small), and then do some post-processing based on user-defined inputs to tidy up. @mick-iqvia has some suggestions related to this, which I'm guessing you'll discuss next week :)

**Comment by chrisknoll** [comment]
> Would we not want A + B -> B in this case?

Yes, and my understanding of how this would resolve is:

| Cohort  | Start | End |
| ------------- | ------------- | ------------- | 
| A  | 1/1  | 1/1 |
| B  | 1/1  | 1/1 |
| B  | 1/1  | 1/8 |

And because a and b have the exact same start/end, we get A+B -> B

I need to think more of your suggestion there, but I think we get into some problem space when we have to think about 'inclusive end dates' vs non-inclusive...

**Comment by chrisknoll** [comment]
Per our Conference Call on 2/26:  we're going to adjust the algorithm to instead of 'collapsing dates' we decided to go with 'minimum span length' where we will drop any segment that is < the minimum span length.

For simplicity, we are going to use the 'collapseWindow' value as the minimum span length in the algorithm, and separately we will rename the fields in the objects to reflect 'minSpanLength' in the Entity/Tables/paramaters in a new PR (after this has been merged with master).

**Comment by chrisknoll** [comment]
Hi @hms1 , the pathway analsysis query has been updated in master since this PR, but the conflicting lines is just realted to some CAST operations that were required for some SQL dialects. Would you mind syncing your upstream with the latest master branch in WebAPI, then resolve conflicts?

Otherwise, what we can do is take this PR into a side-branch on the OHDSI repository, resolve the conflicts there, and make another PR with your updatse that will be based of an OHDSI/WebAPI branch.

**Comment by chrisknoll** [comment]
Hi, @hms1. I don't think we'll be able to fold this in for the 2.7 release.  We're pretty far over-due with making this release, we'd still want to do extensive testing on the changes, and I think it is best that we not overload the 'collapse window' value with the 'minimum span length'. To handle the new paramater, we really should think about updating the object model to add the new parameter and alter the database tables to store it.  So, this would better be a 2.8 item.  Sorry for the inconvenience.

**Comment by hms1** [comment]
> Hi, @hms1. I don't think we'll be able to fold this in for the 2.7 release. We're pretty far over-due with making this release, we'd still want to do extensive testing on the changes, and I think it is best that we not overload the 'collapse window' value with the 'minimum span length'. To handle the new paramater, we really should think about updating the object model to add the new parameter and alter the database tables to store it. So, this would better be a 2.8 item. Sorry for the inconvenience.

**Comment by hms1** [comment]
No problem at all... when is 2.8 planned for? 

I've been working with the Pathways scripts quite a lot over the past few weeks, and have a couple of additional suggestions that could be considered if there's time.

**Comment by chrisknoll** [comment]
2.8 should be sometime end of April.  Sorry I did not reply to this sooner.  Did you mean to close this PR?  I was hoping we could start working on incorporating these changes.

**Comment by hms1** [comment]
Sorry, must have closed it by accident.

Happy to start working on this again if there's anything else you need.

**Comment by pavgra** [comment]
@chrisknoll , there are conflicts in PR. Could you resolve?

**Comment by chrisknoll** [comment]
It's in a fork, so I can't commit anything to it. However, I wanted to create a new branch that incorporates the proposed changes here, allows the existing functionality of collapse window, and introduces the new paramater of 'minSegmentLength' described here: https://github.com/OHDSI/StandardizedAnalysisAPI/pull/23 and here: https://github.com/OHDSI/WebAPI/pull/1049. Unfortunately, my attention has been very distracted with other issues and I haven't been able to find time to drill into this work and complete it. I plan on making this my solitary focus int he coming week.

**Comment by anthonysena** [comment]
Linking back to #779.

**Comment by chrisknoll** [comment]
Will use #1211 for this fix.

**Pull request closed without merging**